File: Tradio.Client/src/app/core/services/notification.service.ts
import { Injectable } from '@angular/core';
import { HubConnection, HubConnectionBuilder } from '@microsoft/signalr';
import { Subject } from 'rxjs/internal/Subject';
import { MessageFromSingalRModel } from '../responses/message-from-singal-r.model';

@Injectable({
  providedIn: 'root',
})
export class NotificationService {
  private hubConnection!: HubConnection;
  private messageSubject = new Subject<MessageFromSingalRModel>();
  messages$ = this.messageSubject.asObservable();

  async start(userId: string) {
    this.hubConnection = new HubConnectionBuilder()
      .withUrl('https://localhost:7171/chatHub')
      .withAutomaticReconnect()
      .build();

    this.hubConnection.on(
      'ReceiveMessage',
      (message: MessageFromSingalRModel) => {
        this.messageSubject.next(message);
      },
    );

    await this.hubConnection.start();
    await this.joinUser(userId);
  }

  async joinToChats(chatIds: number[]) {
    for (let chatId of chatIds) {
      await this.hubConnection.invoke('AddToChatAsync', chatId);
    }
  }

  private async joinUser(userId: string) {
    await this.hubConnection.invoke('AddUserAsync', userId);
  }
}

File: Tradio.Mobile/app/src/main/java/com/example/tradio/data/remote/source/NotificationService.kt
package com.example.tradio.data.remote.source

import android.util.Log
import com.example.tradio.data.remote.dto.responses.MessageFromSignalRModel
import com.microsoft.signalr.HubConnection
import com.microsoft.signalr.HubConnectionBuilder
import com.microsoft.signalr.HubConnectionState
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.asSharedFlow

class NotificationService {

    private var hubConnection: HubConnection? = null

    private val _messages = MutableSharedFlow<MessageFromSignalRModel>(extraBufferCapacity = 64)
    val messages = _messages.asSharedFlow()

    private val hubUrl = "http://10.0.2.2:5188/chatHub"

    fun start(userId: String, onConnected: () -> Unit = {}) {
        if (hubConnection?.connectionState == HubConnectionState.CONNECTED) {
            onConnected()
            return
        }

        hubConnection = HubConnectionBuilder
            .create(hubUrl)
            .build()

        hubConnection?.on("ReceiveMessage", { message ->
            Log.d("SignalR", "Received object: $message")
            _messages.tryEmit(message)
        }, MessageFromSignalRModel::class.java)

        try {
            hubConnection?.start()?.blockingAwait()
            Log.d("SignalR", "Connection established successfully")

            joinUser(userId)
            onConnected()
        } catch (e: Exception) {
            Log.e("SignalR", "SignalR start error: ${e.localizedMessage}")
        }
    }

    fun joinToChats(chatIds: List<Int>) {
        if (hubConnection?.connectionState == HubConnectionState.CONNECTED) {
            for (chatId in chatIds) {
                Log.d("SignalR", "Sending AddToChatAsync for chat ID: $chatId")
                hubConnection?.send("AddToChatAsync", chatId)
            }
        } else {
            Log.e("SignalR", "Cannot join chats: Hub is not connected!")
        }
    }

    private fun joinUser(userId: String) {
        if (hubConnection?.connectionState == HubConnectionState.CONNECTED) {
            hubConnection?.send("AddUserAsync", userId)
        }
    }

    fun stop() {
        hubConnection?.stop()?.blockingAwait()
    }
}